### PR TITLE
HC-17 and HC-18

### PR DESCRIPTION
**Pull request by Magisus**
This is an unfinished implementation of HC-17 and HC-18. All the tests pass except for the last "can replace an object" tests in config_node_test, which segfault. Any help figuring out why one of the post-replacement object's children is null would be appreciated...

### CODE REVIEW

**Comment by Magisus** [comment]
Never mind, segfault fixed!

**Comment by MikaelSmith** [comment]
\o/

**Comment by Magisus** [comment]
Alright, all classes are implemented and all tests pass. Ready for review!

File: lib/tests/config_node_test.cc [hunk 1]
```diff
@@ -92,4 +94,240 @@ TEST_CASE("simple values", "[config-node]") {
...
+            line_node(10),
+            node_key_value_pair(node_key("baz.abc.ghi"), int_node(52)),
+            line_node(11),
+            node_key_value_pair(node_key("baz.abc.ghi"), int_node(53)),
+            line_node(12),
+            close_brace_node()
+    });
+    REQUIRE(orig_text == original_node->render());
+
+    string final_text = string("foo : bar\nbaz : {\n\t\"abc.def\" : true\n\t") +
```
Comment (MikaelSmith) [inline lib/tests/config_node_test.cc]: Instead of concatenating strings, you can concat string literals, as in
```
"foo : bar\n" "baz : "
"{\n\t\"..."
```
This is C-style string literal concatenation, and can span multiple lines.

File: lib/tests/config_node_test.cc [hunk 2]
```diff
@@ -92,4 +94,240 @@ TEST_CASE("simple values", "[config-node]") {
...
+            string("//this is a comment about the below setting") +
+            string("\n\n\tabc : {\n\t\tdef : false\n\t\t\n\t\t\"this.does.not.exist@@@+$#\"") +
+            string(" : {\n\t\t  end : doesnotexist\n\t\t}\n\t}\n}\n\nbaz.abc.ghi : randomunquotedString\n}");
```
Comment (MikaelSmith) [inline lib/tests/config_node_test.cc]: That last `}` is weird, but definitely matches the original test.

**Comment by MikaelSmith** [comment]
Any notes you could add to describe what each class is would be helpful. Not sure what a field is.

File: lib/src/nodes/abstract_config_node.cc [hunk 3]
```diff
@@ -1,4 +1,6 @@
 #include <internal/nodes/abstract_config_node.hpp>
+#include <iostream>
```
Comment (MikaelSmith) [inline lib/src/nodes/abstract_config_node.cc]: Why were these needed?

File: lib/src/nodes/config_node_object.cc [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+
+    bool config_node_object::has_value(path desired_path) const {
+        for (auto&& node : children()) {
+            if (auto field = dynamic_pointer_cast<config_node_field>(node)) {
+                path key = field->path()->get_path();
+                if (key == desired_path || key.starts_with(desired_path)) {
+                    return true;
+                } else if (desired_path.starts_with(key)) {
+                    if (auto object = dynamic_pointer_cast<config_node_object>(field->get_value())) {
+                        path remaning_path = desired_path.sub_path(key.length());
```
Comment (MikaelSmith) [inline lib/src/nodes/config_node_object.cc]: Typo in 'remaning_path'.

File: lib/src/nodes/config_node_object.cc [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+                    }
+                }
+            }
+        }
+        return false;
+    }
+
+    // TODO: Just gonna throw this out here, this method is disgustingly long and hard to read...
+    // (line for line port from Java)
+    shared_ptr<config_node_object> config_node_object::change_value_on_path(path desired_path, shared_node_value value,
```
Comment (MikaelSmith) [inline lib/src/nodes/config_node_object.cc]: Candidate for refactoring if we have to make significant changes to it (bugs, new features). Probably not worth it until then, because we'd also want to change the other projects.

**Comment by Magisus** [comment]
I only have a vague idea of what the classes represent myself. There is no description in the original code. But I'll give it a shot based on the tests.

File: lib/src/path.cc [hunk 6]
```diff
@@ -112,13 +112,17 @@ namespace hocon {
             count--;
             builder.append_key(*from.first());
             from = from.remainder();
-            if (from.has_remainder()) {
+            if (from.empty()) {
```
Comment (MikaelSmith) [inline lib/src/path.cc]: Wouldn't this be the same as `from == path()`?

File: lib/src/path.cc [hunk 7]
```diff
@@ -112,13 +112,17 @@ namespace hocon {
...
                 throw config_exception("sub_path last_index out of range");
             }
         }
         return builder.result();
     }
 
+    bool path::empty() const {
+        return first() == nullptr;
```
Comment (MikaelSmith) [inline lib/src/path.cc]: This seems like it could just return __path.isEmpty().

File: lib/src/path_parser.cc [hunk 8]
```diff
@@ -143,14 +143,18 @@ namespace hocon {
             return token_list { t };
         }
 
+        // The split iterator tacks an empty string onto the end of the iterator if the string
+        // ends in the split character; so we need to remove trailing '.'
+        string text_to_split = boost::trim_right_copy_if(token_text, is_any_of("."));
```
Comment (MikaelSmith) [inline lib/src/path_parser.cc]: Line 165 seems like it duplicates this. Why didn't it work? (also it could've used `token_text.back()`).

Comment (MikaelSmith) [inline lib/src/path_parser.cc]: And if we don't need that code, we could just drop the const on token_text and trim it in-place.

File: lib/src/path_parser.cc [hunk 9]
```diff
@@ -143,14 +143,18 @@ namespace hocon {
...
+
         auto token_it = boost::make_iterator_range(make_split_iterator(
-                token_text, token_finder(is_any_of("."))), split_iterator<string::const_iterator>());
+                text_to_split, token_finder(is_any_of("."))), split_iterator<string::iterator>());
 
         token_list split_tokens;
         for (auto& s : token_it) {
             string token_string { s.begin(), s.end() };
```
Comment (MikaelSmith) [inline lib/src/path_parser.cc]: Is this really needed? I'd think s would implicitly convert to string as needed for the make_shared calls.

File: lib/src/path_parser.cc (hunk 8 above)
Comment (Magisus) [inline lib/src/path_parser.cc]: So these address two separate issues: line 165 handles the case where token_text is e.g. "a.b.c", in which c is considered a key to a final value, I believe (not 100% sure, but that's kind of what the tests look like). My problem was in the case of "a.b.c.", where the split iterator returned an extra empty string at the end, resulting in double "." when using the tokens to make a path.

File: lib/src/path_parser.cc (hunk 9 above)
Comment (Magisus) [inline lib/src/path_parser.cc]: Apparently it doesn't, I get a compile error.

File: lib/src/path_parser.cc (hunk 8 above)
Comment (MikaelSmith) [inline lib/src/path_parser.cc]: Ah, I misunderstood what it was checking (`!=`, not `==`). Also double checked Java's [string.split](http://docs.oracle.com/javase/7/docs/api/java/lang/String.html#split(java.lang.String)) implementation to verify it leaves off trailing empty strings.

Comment (Magisus) [inline lib/src/path_parser.cc]: I'll put in a second comment to that effect.

Comment (MikaelSmith) [inline lib/src/path_parser.cc]: We could avoid copying the string by doing the `!= '.'` before-hand and saving it as a boolean.

**Comment by MikaelSmith** [comment]
This looks reasonable. :+1:
